File: src/Rulesage.Shared/Repositories/Abstractions/IRuleRepository.cs
using Rulesage.Common.Grammar.Ast;

namespace Rulesage.Shared.Repositories.Abstractions;

public interface IRuleRepository : IDocumentRepository
{
    Task<IEnumerable<RuleExpr>> FindByIdsAsync(IEnumerable<string> ids, CancellationToken cancellationToken = default);

    Task<IEnumerable<(RuleExpr, float)>> FindOrderByCosineDistanceAsync(string community, float[] queryVector, int skip, int take,
        CancellationToken cancellationToken = default);
    
    Task<bool> SaveAsync(IEnumerable<RuleExpr> rules, CancellationToken cancellationToken = default);
}

File: src/Rulesage.Common/Repositories/Abstractions/IRuleRepository.fs
namespace Rulesage.Common.Repositories.Abstractions

open System.Runtime.InteropServices
open System.Threading
open System.Threading.Tasks
open Rulesage.Common.Grammar.Ast

type IRuleRepository =
    inherit IDocumentRepository

    abstract FindByIdsAsync:
        ids: string seq * [<Optional; DefaultParameterValue(CancellationToken())>] cancellationToken: CancellationToken ->
            Task<RuleExpr seq>

    abstract FindOrderByCosineDistanceAsync:
        contextCommunity: string *
        queryVector: float32[] *
        skip: int *
        take: int *
        [<Optional; DefaultParameterValue(CancellationToken())>] cancellationToken: CancellationToken ->
            Task<(RuleExpr * float32) seq>

    abstract SaveAsync:
        rules: RuleExpr seq *
        [<Optional; DefaultParameterValue(CancellationToken())>] cancellationToken: CancellationToken ->
            Task<int>
